src/AccessLogs.sol

```diff
@@ -0,0 +1,190 @@
+// specific language governing permissions and limitations under the License.
+
+/// @dev This file is generated from helper_scripts/generate_AccessLogs.sh, one should not modify the content directly
+
+pragma solidity ^0.8.0;
+
+import "./Memory.sol";
+import "./UArchConstants.sol";
+
+library AccessLogs {
+    using AccessLogs for bytes;
+    using Memory for Memory.AlignedSize;
+
+    struct Context {
+        bytes32 currentRootHash;
+        bytes buffer;
+        uint128 pointer;
+    }
+
+    /// @notice Swap byte order of unsigned ints with 64 bytes
+    /// @param num number to have bytes swapped
+    function uint64SwapEndian(uint64 num) internal pure returns (uint64) {
+        uint64 output = ((num & 0x00000000000000ff) << 56)
+            | ((num & 0x000000000000ff00) << 40)
+            | ((num & 0x0000000000ff0000) << 24) | ((num & 0x00000000ff000000) << 8)
+            | ((num & 0x000000ff00000000) >> 8) | ((num & 0x0000ff0000000000) >> 24)
+            | ((num & 0x00ff000000000000) >> 40)
+            | ((num & 0xff00000000000000) >> 56);
+
+        return output;
+    }
+
+    function writeBytes32(bytes memory data, uint128 offset, bytes32 val)
+        internal
+        pure
+    {
+        assembly {
+            mstore(add(data, add(offset, 32)), val)
+        }
+    }
+
+    function toBytes32(bytes memory data, uint128 offset)
+        internal
+        pure
+        returns (bytes32)
+    {
+        bytes32 temp;
+        // Get 32 bytes from data
+        assembly {
+            temp := mload(add(data, add(offset, 32)))
+        }
+        return temp;
+    }
+
+    /// @dev bytes buffer layouts differently for `readWord` and `writeWord`,
+    ///`readWord` [8 bytes as uint64 value, 32 bytes as drive hash, 61 * 32 bytes as sibling proofs]
+    ///`writeWord` [32 bytes as old drive hash, 32 * 61 bytes as sibling proofs]
+
+    //
+    // Read methods
+    //
+    function readRegion(
+        AccessLogs.Context memory a,
+        Memory.Region memory region
+    ) internal pure returns (bytes32) {
+        bytes32 drive = a.buffer.toBytes32(a.pointer);
+        a.pointer += 32;
+        (bytes32 rootHash, uint8 siblingCount) =
+            getRoot(region, drive, a.buffer, a.pointer);
+        a.pointer += uint128(siblingCount) * 32;
+
+        require(a.currentRootHash == rootHash, "Read region root doesn't match");
+
+        return drive;
+    }
+
+    function readLeaf(AccessLogs.Context memory a, Memory.Stride readStride)
+        internal
+        pure
+        returns (bytes32)
+    {
+        Memory.Region memory r =
+            Memory.regionFromStride(readStride, Memory.alignedSizeFromLog2(0));
+        return readRegion(a, r);
+    }
+
+    function readWord(
+        AccessLogs.Context memory a,
+        Memory.PhysicalAddress readAddress
+    ) internal pure returns (uint64) {
+        uint64 val =
+            uint64SwapEndian(uint64(bytes8(a.buffer.toBytes32(a.pointer))));
+        a.pointer += 8;
```

Comment:
Likewise for these two lines: they could be a single `consumeWord` method on `Buffer` type.